src/resolve_addr.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+
+use futures::Async;
+use futures::Future;
+use futures::Poll;
+use std::error::Error;
+use std::fmt;
+use std::net::SocketAddr;
+use std::net::ToSocketAddrs;
+
+/// Go-style network address parsing. Returns a future.
+/// Examples:
+/// "192.0.2.1:25"
+/// ":80"
+/// "[2001:db8::1]:80"
+/// "198.51.100.1:80"
+/// "deno.land:443"
+pub fn resolve_addr(address: &str) -> ResolveAddrFuture {
+  ResolveAddrFuture {
+    address: address.to_string(),
+  }
+}
+
+#[derive(Debug)]
+pub enum ResolveAddrError {
+  Syntax,
+  Resolution(std::io::Error),
+}
+
+impl fmt::Display for ResolveAddrError {
+  fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+    fmt.write_str(self.description())
+  }
+}
+
+impl Error for ResolveAddrError {
+  fn description(&self) -> &str {
+    match self {
+      ResolveAddrError::Syntax => "invalid address syntax",
+      ResolveAddrError::Resolution(e) => e.description(),
+    }
+  }
+}
+
+pub struct ResolveAddrFuture {
+  address: String,
+}
+
+impl Future for ResolveAddrFuture {
+  type Item = SocketAddr;
+  type Error = ResolveAddrError;
+
+  fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    // The implementation of this is not actually async at the moment,
+    // however we intend to use async DNS resolution in the future and
+    // so we expose this as a future instead of Result.
+    match split(&self.address) {
+      None => Err(ResolveAddrError::Syntax),
+      Some(addr_port_pair) => {
+        // I absolutely despise the .to_socket_addrs() API.
+        let r = addr_port_pair
+          .to_socket_addrs()
+          .map_err(|e| ResolveAddrError::Resolution(e));
+
+        r.and_then(|mut iter| match iter.next() {
+          Some(a) => Ok(Async::Ready(a)),
+          None => panic!("There should be at least one result"),
+        })
+      }
+    }
+  }
+}
+
+fn split<'a>(address: &'a str) -> Option<(&'a str, u16)> {
+  address.rfind(":").and_then(|i| {
+    let (a, p) = address.split_at(i);
+    // Default to localhost if given just the port. Example: ":80"
+    let addr = if a.len() > 0 { a } else { "127.0.0.1" };
```

Comment:
Does go also default to IPv4 and localhost? Other valid defaults might be `::1` and `0.0.0.0`.

Reply:
https://github.com/golang/go/blob/7cbfa55b5d17c8deaecff05e4221f828467cfa97/src/net/ipsock.go#L275-L277

It appears to default to "0.0.0.0". I've updated the code. Good catch.